### PR TITLE
Change confirmation dialog to toasts

### PR DESCRIPTION
**Pull request by lw64**
Fixes https://github.com/sonnyp/Workbench/issues/84.

### CODE REVIEW

File: src/util.js [hunk 1]
```diff
@@ -148,6 +148,9 @@ export function disconnect_signals(target, handler_ids) {
 }
 
 export function replaceBufferText(buffer, text, scroll_start = true) {
+  if (text == "") {
+    text = " ";
```
Comment (sonnyp) [inline src/util.js:152]: a bit hackish but fine by me - should we report an issue upstream?

File: src/window.js [hunk 2]
```diff
@@ -218,46 +220,48 @@ export default function Window({ application }) {
...
+            //     <property name="label">
+            //       Edit Style and UI to reload the Preview
+            //     </property>
+            //     <property name="justify">center</property>
+            //   </object>
+            // </child>
+          });
         });
-      });
+      }
```
Comment (sonnyp) [inline src/window.js:264]:
```suggestion
      }
```

File: src/window.js [hunk 3]
```diff
@@ -389,25 +415,27 @@ export default function Window({ application }) {
...
   const text_decoder = new TextDecoder();
   async function openFile(file) {
     const language = getLanguageForFile(file);
-    if (!language) return;
-
-    const agreed = await confirmDiscard();
-    if (!agreed) return;
+    if (!language) {
+      const toast = new Adw.Toast({
+        title: "File is not loadable",
```
Comment (sonnyp) [inline src/window.js:423]:
```suggestion
        title: _("Unknown file type"),
```

File: src/window.js [hunk 4]
```diff
@@ -389,25 +415,27 @@ export default function Window({ application }) {
...
+      });
+      toast_overlay.add_toast(toast);
+      return;
+    }
+    
+    const toast = new Adw.Toast({
+      title: "Opened File",
```
Comment (sonnyp) [inline src/window.js:430]:
```suggestion
      title: _("File opened"),
```

File: src/window.js [hunk 5]
```diff
@@ -389,25 +415,27 @@ export default function Window({ application }) {
...
+      button_label: "Undo",
```
Comment (sonnyp) [inline src/window.js:431]:
```suggestion
      button_label: _("Undo"_,
```

**Review by sonnyp — CHANGES_REQUESTED** [review]
Looking good, one request and one comment

---

I feel like we should restore the previous panel configuration on undo or it can be a bit confusing.

So for example:

panels a and c are open
opens demo which closes a and c but open b and d
undo should restore previous panels (only a and c visible)

I think we can do this easily by passing the panel configuration to the undo action.

---

Not super important but should we show the toast if the previous buffer was empty? (or `text.trim() === ''`)

File: src/window.js (hunk 3 above)
Comment (sonnyp) [reply src/window.js:423]: @bertob wdyt?

File: src/window.js (hunk 4 above)
Comment (sonnyp) [reply src/window.js:430]: @bertob wdyt?

File: src/window.js (hunk 3 above)
Comment (bertob) [reply src/window.js:423]: Sounds a bit odd, maybe "This file can not be loaded"?

Comment (sonnyp) [reply src/window.js:423]: "This file can not be loaded" sounds good to me

File: src/window.js (hunk 4 above)
Comment (sonnyp) [reply src/window.js:430]: i guess "The file has been loaded" I guess then

File: src/window.js [hunk 6]
```diff
@@ -326,16 +329,56 @@ export default function Window({ application }) {
...
       replaceBufferText(buffer, str);
     }
 
-    const { javascript, css, xml, blueprint, vala, panels } =
-      readDemo(demo_name);
+    const { javascript, css, xml, blueprint, vala, panels }
+       = readDemo(demo_name);
+    
+    const toast = new Adw.Toast({
+      title: "Loaded Demo",
```
Comment (sonnyp) [inline src/window.js:364]:
```suggestion
      title: _("The demo has been loaded")
```

File: src/window.js [hunk 7]
```diff
@@ -326,16 +329,56 @@ export default function Window({ application }) {
...
+      button_label: "Undo",
```
Comment (sonnyp) [inline src/window.js:365]:
```suggestion
      button_label: _("Undo"),
```

File: src/window.js [hunk 8]
```diff
@@ -389,25 +430,36 @@ export default function Window({ application }) {
...
   const text_decoder = new TextDecoder();
   async function openFile(file) {
     const language = getLanguageForFile(file);
-    if (!language) return;
-
-    const agreed = await confirmDiscard();
-    if (!agreed) return;
+    if (!language) {
+      const toast = new Adw.Toast({
+        title: _("This file can not be loaded"),
```
Comment (sonnyp) [inline src/window.js:438]:
```suggestion
        title: _("This file cannot be loaded"),
```

File: src/Document.js [hunk 9]
```diff
@@ -26,8 +26,7 @@ export default function Document({
 
   loadSourceBuffer({ file: source_file, buffer })
     .then((success) => {
-      if (!success) replaceBufferText(buffer, placeholder, true);
-      settings.set_boolean("has-edits", false);
+      if (!success) buffer.set_text(placeholder, -1);
```
Comment (sonnyp) [inline src/Document.js:29]: accident?

**Review by sonnyp — CHANGES_REQUESTED** [review]
Please solve conlicts :smile:

**Review by sonnyp — APPROVED** [review]
:partying_face:

**Pull request merged**